modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj

```diff
@@ -231,6 +240,45 @@
   ;; Work around this by converting the timestamps to minutes instead before calling DATEADD().
   (date-add :minute (hx// expr 60) (hx/literal "1970-01-01")))
 
+(defonce
+  ^{:private true
+    :doc     "A map of all zone-id to the corresponding window-zone.
+             I.e {\"Asia/Tokyo\" \"Tokyo Standard Time\"}"}
+  zone-id->windows-zone
+  (let [data (-> (io/resource "timezones/windowsZones.xml")
+                 io/reader
+                 xml/parse
+                 :content
+                 second
+                 :content
+                 first
+                 :content)]
+    (->> (for [mapZone data
+               :let [attrs       (:attrs mapZone)
+                     window-zone (:other attrs)
+                     zone-ids    (str/split (:type attrs) #" ")]]
+           (zipmap zone-ids (repeat window-zone)))
+         (cons {"UTC""UTC"})
+         (apply merge))))
+
+(defmethod sql.qp/->honeysql [:sqlserver :convert-timezone]
+  [driver [_ arg target-timezone source-timezone]]
+  (let [expr            (sql.qp/->honeysql driver arg)
+        datetimeoffset? (hx/is-of-type? expr "datetimeoffset")]
+    (when (and datetimeoffset? source-timezone)
+      (throw (ex-info (tru "`datetimeoffset` columns shouldn''t have a `source timezone`")
+                    {:type            qp.error-type/invalid-parameter
+                     :target-timezone target-timezone
+                     :source-timezone source-timezone})))
+    (let [source-timezone (or source-timezone (qp.timezone/results-timezone-id))]
+      (cond-> expr
+        (and (not datetimeoffset?) source-timezone)
+        (hx/->AtTimeZone (zone-id->windows-zone source-timezone))
+        target-timezone
+        (hx/->AtTimeZone (zone-id->windows-zone target-timezone))
+        true
+        hx/->datetime))))
```

Comment:
Same as before, `source-timezone` and `target-timezone` can never be falsy.

Reply:
updated here ca903dfe8b